crates/bevy_ecs/src/component.rs

```diff
@@ -143,7 +178,18 @@ impl ComponentDescriptor {
         x.cast::<T>().drop_in_place()
     }
 
-    pub fn new<T: Component>(storage_type: StorageType) -> Self {
+    pub fn new<T: Component>() -> Self {
+        Self {
+            name: std::any::type_name::<T>().to_string(),
+            storage_type: T::Storage::STORAGE_TYPE,
```

Comment:
We should probably add a SAFETY line above the `storage_type` field in the struct definition (like we do for `is_send_and_sync`). We can't let people override this / set it to something mismatched.